### DISCUSSION TITLE
How to implement CubeSat Space Protocol (CSP)

### DISCUSSION THREAD

**Discussion by ghaith99** [yamcs/yamcs#670] [Q&A]
Hello, 

I'm trying assess the feasibility of integrating CubeSat Space Protocol (CSP) for a Cubesat project with Yamcs as a Mission Operations and Ground Support system.

I would appreciate if someone can guide me on which part of the documentation I should look into. 

Thanks in advance

**Comment by xpromache** [comment]
Please read this page:
https://docs.yamcs.org/yamcs-server-manual/links/packet-preprocessor/

I think you would need to implement a specific CSP packet-preprocessor but
I don't know enough about CSP to offer more hints.

**Comment by shresthdeepgupta** [comment]
I'm also trying to build a CSP adaptor for commands sent from Yamcs to CubeSat. Any leads in the direction will help a lot.

**Comment by xpromache** [comment]
As far as I know, CSP is a fairly simple structure which should be easy to
define in the Yamcs MDB. There is the question of Preprocessor and
Postprocessor - not clear to me if the generic ones would work, I think
they should.

Can you contribute a simple python simulator that implements CSP to some
basic level (i.e generate some 1Hz TM and accept a command or two)?
Then I could add an example Yamcs configuration that works with that
simulator.

**Comment by gokulapixxel** [comment]
Are there any implementations for this?

**Reply by xpromache** [reply]
As I said above, if someone contributes a simple CSP simulator, I will look into making an example in Yamcs to control it.

**Comment by fqqb** [comment]
As of Yamcs v5.9.x, Yamcs comes with some built-in CSP 1.x pre/post processors:

* ``org.yamcs.tctm.csp.CspPacketPreprocessor``
* ``org.yamcs.tctm.csp.CspCommandPostprocessor``

They support the CRC flag too.

Then this would need to be combined with an appropriate MDB definition. There's an example of that here: https://github.com/yamcs/pymdb/blob/master/examples/csp.xml
